# Dataset assembly and analysis in palaeognathous birds

## Setup

```{r}
setwd("/Users/chadeliason/Documents/UT/projects/phenome")

# load packages
devtools::load_all('~/github/nexustools')

library(Claddis)
library(dplyr)
library(igraph)
library(readxl)
library(gridExtra)
library(rwty)
library(phangorn)

# set options
# options(dev = "pdf")
# opts_knit$set(progress = TRUE, verbose = TRUE, dev = 'pdf')
```

## Read in a set of nexus files

```{r}
files <- list.files('~/Documents/UT/projects/phenome/data', pattern='\\.nex', full.names = TRUE)

system.time(nexs <- lapply(files, read.nex))

# system.time(nexs2 <- lapply(files, ReadMorphNexus))

# 3 seconds for my function, 25+ seconds for Claddis ReaDMorphNexus function
# this is because Claddis is reading in data as factors and I'm using character data
# also - Claddis returns an error for some of the datasets...
```


## Convert supraspecific terminals in datasets to species (based on specimens evaluated)

```{r}
nexs[[grep("bertelli_2014", files)]] <- collapse.nex(nexs[[grep("bertelli_2014", files)]],
  map = list("Lithornis_spp" = c("Lithornis_celetius", "Lithornis_vulturinus")))

nexs[[grep("bertelli_chiappe", files)]] <- collapse.nex(nexs[[grep("bertelli_chiappe", files)]],
  map = list("Lithornis" = c("Lithornis_hookeri", "Lithornis_vulturinus", "Lithornis_nasi", "Lithornis_plebius")))

nexs[[grep("bledsoe", files)]] <- collapse.nex(nexs[[grep("bledsoe", files)]],
  map = list('Ancestor' = c('Tinamus_solitarius', 'Crypturellus_undulatus', 'Rhynchotus_rufescens', 'Nothoprocta_perdicaria', 'Nothoprocta_cinerascens', 'Nothura_darwinii', 'Nothura_maculosa', 'Eudromia_elegans'),
    'Struthio' = 'Struthio_camelus', 'Rhea' = 'Rhea_americana',
    'Dromaius' = 'Dromaius_novaehollandiae',
    'Apteryx' = c('Apteryx_australis','Apteryx_owenii'),
    'Dinornithidae' = c('Pachyornis_elephantopus', 'Euryapteryx_sp.', 'Dinornis_maximus', 'Dinornis_gracilis', 'Dinornis_struthoides')))

# Bourdon et al. (2009) - 20 ordered  multistate characters (6, 7, 14, 16, 18, 21, 28, 34, 43, 44, 70, 71, 87, 88, 93, 103, 109, 114, 117, and 122) 
nexs[[grep("bourdon", files)]] <- collapse.nex(nexs[[grep("bourdon", files)]],
  map = list('Apteryx_spp' = 'Apteryx_haastii',
    'Tinamidae_' = c('Rhynchotus_rufescens', 'Eudromia_elegans')))
nexs[[grep("bourdon", files)]]$taxlabels[grep("Emeus", nexs[[grep("bourdon", files)]]$tax)] <- "Emeus_crassus"

nexs[[grep("johnston", files)]] <- collapse.nex(nexs[[grep("johnston", files)]],
  map = list('Casuarius_spp' = 'Casuarius_casuarius',
    'Apteryx_spp' = c('Apteryx_mantelli', 'Apteryx_australis'),
    'Tinamidae_' = c('Tinamus_major', 'Eudromia_elegans'), 'Dinornithiformes' = c('Megalapteryx_didinus', 'Anomalopteryx_didiformis', 'Euryapteryx_gravis', 'Pachyornis_geranoides', 'Dinornis_novaezealandiae'),
    'Rhea_americana' = c('Rhea_americana', 'Rhea_pennata')))

nexs[[grep("lee_1997", files)]] <- collapse.nex(nexs[[grep("lee_1997", files)]],
  map = list('Tinamidae' = c('Tinamus_tao', 'Tinamus_major', 'Crypturellus_cinnamomeus', 'Crypturellus_undulatus', 'Crypturellus_noctivagus', 'Nothoprocta_cinerascens', 'Eudromia_elegans', 'Rhynchotus_rufescens'),
    'Dinornithidae' = c('Megalapteryx_didinus', 'Emeus_crassus', 'Dinornis_maximus', 'Dinornis_robustus', 'Pachyornis_elephantopus', 'Euryapteryx_sp.'),
    'Struthio' = 'Struthio_camelus',
    'Rhea' = 'Rhea_americana',
    'Casuarius' = c('Casuarius_casuarius', 'Casuarius_unappendiculatus'),
    'Dromaius' = 'Dromaius_novaehollandiae',
    'Apteryx' = 'Apteryx_australis',
    'Galloanseres' = c('Megapodius_freycinet', 'Anhima_cornuta', 'Leipoa_ocellata', 'Crax_mitu', 'Nothocrax_urumutum')))

nexs[[grep("livezey", files)]] <- collapse.nex(nexs[[grep("livezey", files)]],
  map = list('Chauna' = 'Chauna_torquata',
    'Anas' = 'Anas_platyrhynchos',
    'Gallus' = 'Gallus_gallus'))

nexs[[grep("mayr_2011", files)]] <- collapse.nex(nexs[[grep("mayr_2011", files)]],
  map = list('Rheidae' = 'Rhea',
    'Apterygidae' = 'Apteryx',
    'Tinamidae' = c('Crypturellus_cinnamomeus', 'Nothura', 'Rhynchotus_rufescens', 'Tinamus_solitarius'),
    'Galliformes' = 'Gallus',
    'Anhimidae' = 'Chauna_torquata',
    'Anatidae' = c('Anas', 'Anser')))

nexs[[grep("mckitrick", files)]] <- collapse.nex(nexs[[grep("mckitrick", files)]],
  map = list('Chauna' = 'Chauna_torquata',
    'Anaspiat' = 'Anas_platyrhynchos',
    'Apteryx_spp' = 'Apteryx_mantelli',
    'Tinamotis_spp' = c('Tinamotis_ingoufi', 'Tinamotis_pentlandii'),
    'Tinamus_spp' = c('Tinamus_guttatus', 'Tinamus_major', 'Tinamus_osgoodi', 'Tinamus_tao', 'Tinamus_solitarius'),
    'Casuarius_spp' = c('Casuarius_bennetti', 'Casuarius_casuarius'),
    'Nothura_spp' = 'Nothura_maculosa',
    'Eudromia_spp' = 'Eudromia_elegans',
    'Dromaius_spp' = 'Dromaius_novaehollandiae',
    'Crypturellus_spp' = 'Crypturellus_cinnamomeus'))

# remove bad character
nexs[[grep("mckitrick", files)]] <- nexs[[grep("mckitrick", files)]][, !grepl("Neognath monophyly", nexs[[grep("mckitrick", files)]]$charlab)]


# rename
# nexs[[grep("nesbitt", files)]]$taxlabels[grep("Lithornis_celetius_combined", nexs[[grep("nesbitt", files)]]$tax)] <- "Lithornis_celetius"

# nexs[[grep("nesbitt", files)]]$taxlabels[grep("Lithornis_promiscuus_combined", nexs[[grep("nesbitt", files)]]$tax)] <- "Lithornis_promiscuus"

nexs[[grep("zelenitsky", files)]] <- collapse.nex(nexs[[grep("zelenitsky", files)]], map = list('Apteryx_australis_mantelli' = 'Apteryx_mantelli'))
```


## Concatenate nexus files

```{r}
# list of all taxa in input files
taxa <- sort(unique(unlist(sapply(nexs, '[[', 'taxlabels'))))
taxa <- gsub('\\d', '', taxa)

# taxa to include in concatenated nexus file
tax <- grep('Aepyorn|Anomalopt|Aptery|Casuar|Cryptur|Dinorn|Dromaius|Emeus|Eudromia|Euryapteryx|Lithorn|Megalapteryx|Mullerornis|Nothocercus|Nothoprocta|Nothura|Pachyornis|Palaeotis|Megapod|Pterocnemia|Rhea|Rhynchotus|Struthio_camelus|Taoniscus|Tinamotis|Tinamus|Gallus|Anas|Anatidae|Chauna|Megapodi|Pseudocrypturus|Paracathartes|Ichthyornis', taxa, value=T)

allnex <- merge.nex(nexs, taxa = tax)

# replace _spp with genus name
nms <- allnex$tax[grep("_sp[p]*", allnex$tax)]
fixes <- str_extract(nms, "[A-Z][a-z]+")
names(fixes) <- nms
map <- as.list(fixes)

allnex <- collapse.nex(allnex, map = map, method = "merge")

allnex <- collapse.nex(allnex,
  map = list('Pterocnemia_pennata' = 'Rhea_pennata',
    'Rhynchotus_rufescens_macullocollis' = 'Rhynchotus_rufescens',
    'Rhynchotus_rufescens_pallescens' = 'Rhynchotus_rufescens',
    'Rhynchotus_rufescens_rufescens' = 'Rhynchotus_rufescens',
    'Eudromia_elegans_albida' = 'Eudromia_elegans'), method = 'merge')

# Ichthyornis
# Hesperornis
# Galloanseres (derived ANas + Chauna?, megapods, phasiands, cracids, derive anatids)
# Anhimidae
# Basal neoaves - Metornithies - choose something not totally weird

# remove NAs
allnex <- na.omit(allnex)

allnex
```


## Subset dataset based on species in genomic dataset and important fossils

```{r}
ingenome <- c("Apteryx_haastii", "Apteryx_owenii", "Apteryx_rowi", "Casuarius_casuarius", "Crypturellus_cinnamomeus", "Dromaius_novaehollandiae", "Eudromia_elegans", "Nothoprocta_perdicaria", "Rhea_americana", "Rhea_pennata", "Struthio_camelus", "Tinamus_guttatus")

outgroups <- grep('Anas|Anatidae|Anser|Gallus|Chauna|Megapodius|Ichthyornis', allnex$taxlabels, value=TRUE)

fossils <- grep('Aepyornis|Anomalo|olsoni|reai|paludosa|Dinorni|Emeus|Lithorn|Megalapteryx|Mullerornis|Pachyornis|Palaeotis', allnex$taxlabels, value=TRUE)

ingroups <- subset(allnex, taxlabels %in% c(ingenome, fossils))

# remove invariant characters
allnex2 <- remove.invar(allnex)


# add blank rows for species w/genomic data but missing from phenome
newtax <- setdiff(c(ingenome, "Diogenornis_fragilis", "Opistodactylus_patagonicus"), allnex2$taxlabels)
allnex2$data <- rbind(allnex2$data, matrix(NA, nrow=length(newtax), ncol=ncol(allnex2$data)))
allnex2$taxlabels <- c(allnex2$taxlabels, newtax)

# plot(allnex2, fill="file", legend.pos="top", fsize=6)

# sort and subset the final nexus dataset
allnex.genome <- subset(allnex2, taxlabels %in% c(ingenome, outgroups, fossils, newtax))

# remove "combined" lithornithid taxa from Nesbitt et al. (2015)
allnex.genome <- allnex.genome[-grep("combined", allnex.genome$tax), ]

# allnex.genome <- collapse(allnex.genome, map = list("Lithornis" = "Lithornis_spp", "Dinornis_sp" = "Dinornis_spp", "Dinornithidae" = "Dinornithiformes", "Aepyornis" = "Aepyornis_spp", "Ichthyornis_spp" = "Ichthyornis"))
allnex.genome
```

Make some plots

```{r, eval=FALSE}
plot(na.omit(allnex.genome), fill='file', legend.pos="top")

pdf("figure/allnex2plot.pdf", width=8, height=8)
plot(na.omit(allnex2), fill='file', legend.pos="top", fsize=6)
dev.off()
```


## Sort the dataset taxa

```{r}
rhea <- sort(grep("Rhea", allnex2$tax, value=T))
casuarius <- sort(grep("Casuari", allnex2$tax, value=T))
struthio <- sort(grep("Struthio", allnex2$tax, value=T))
apteryx <- sort(grep("Apteryx", allnex2$tax, value=T))
dinornithids <- sort(grep("Emeus|Dinorn|Euryap|Pachy|Anomalo|Megalap", allnex2$tax, value=T))
aepyornithids <- sort(grep("Muller|Aepyorn", allnex2$tax, value=T))
tinamous <- sort(grep("Rhyncho|Taoniscus|Nothoproc|Cryptur|Eudrom|Tinamus|Tinamotis|Nothocerc|Nothura", allnex2$tax, value=T))
tinamous <- tinamous[-grep("olsoni|paludos", tinamous)]
tinamous <- sort(tinamous)
lithornithids <- sort(grep("Lithorn|Pseudocryp|Paracath", allnex2$tax, value=T))
emus <- sort(grep("Dromaius", allnex2$tax, value=T))

setdiff(allnex2$tax, c(outgroups, rhea,casuarius,struthio,apteryx,dinornithids,aepyornithids,tinamous,lithornithids,emus))

foss <- grep("Diogenornis|Opistodactyl|Palaeotis|Nothura_paludosa|Eudromia_olsoni", allnex2$tax, value=T)

# taxon order: tinamous, kiwi, emu, cassowary, rhea, struthio, fossils, outrgoups

allnex2 <- allnex2[match(c(tinamous,apteryx,emus,casuarius,rhea,struthio,dinornithids,aepyornithids,lithornithids,foss,outgroups), allnex2$tax), ]

allnex.genome <- allnex.genome[na.omit(match(c(tinamous,apteryx,emus,casuarius,rhea,struthio,dinornithids,aepyornithids,lithornithids,foss,outgroups), allnex.genome$tax)), ]

clades <- list(rhea,casuarius,struthio,apteryx,dinornithids,aepyornithids,tinamous,lithornithids,emus,foss,outgroups)
names(clades) <- c("rhea","casuarius","struthio","apteryx","dinornithids","aepyornithids","tinamous","lithornithids","emus","foss","outgroups")
clades <- data.frame(clade=rep(names(clades), times = sapply(clades, length)), spp=unlist(clades))
rownames(clades) <- NULL
```



## Interleave/sort characters based on body region and REGEX searching of trait terms (e.g., humerus)

```{r}
final <- allnex2

tomatch <- str_extract(final$charlab, ".*?(\\,|\\:|$)")
tomatch <- tomatch[final$charpartition %in% c("axial", "pectoral", "forelimb", "pelvic", "hindlimb", "skull")]
names(tomatch) <- which(final$charpartition %in% c("axial", "pectoral", "forelimb", "pelvic", "hindlimb", "skull"))
```

load set of REGEX search terms

```{r}
terms <- read_excel("data/phenome terms.xlsx", sheet=1)
```

find which character labels have a given search term

```{r}
matches <- lapply(terms$"search term", grep, x = tomatch, perl = TRUE)

length(unlist(matches))/length(tomatch)

length(tomatch[-unlist(matches)])  # number of characters not matched
```

get list of non-matches

```{r}
id <- as.numeric(names(tomatch)[-unlist(matches)])
cat(paste0("------\n\n", final$charlabel[id], "\n\n", final$state[id]), sep="\n\n")
# output
cat(tomatch[-unlist(matches)], sep="\n", file="output/unmatchedcharacters.txt")
```

sort by order in matched terms

```{r}
sortid <- sapply(seq_along(matches), function(x) {names(tomatch[matches[[x]]])})
sortid <- as.numeric(unlist(sortid))
sortid <- sortid[!duplicated(sortid)]  # remove dups
```

sort integumentary and other soft-tissue characters by original order in the dataset

```{r}
nistid <- which(!1:ncol(final$data) %in% as.numeric(names(tomatch)))  # nist - non-integumentary characters
```

sort characters at end of group/character partition if they can't be found with REGEX match

```{r}
missid <- which(!1:ncol(final$data) %in% c(sortid, nistid))
```

combine ids and reorder dataset based on these IDs

```{r}
id <- c(sortid,nistid,missid)

# reorder dataset
final <- final[, id]
```


now reorder by character partition

```{r}
final$charpartition <- factor(final$charpartition, levels=c("skull", "hyolingual", "axial", "pectoral", "forelimb", "pelvic", "hindlimb", "integument", "muscle", "eggs", "auditory", "digestive", "respiratory", "pericardium", "chromosomes", "urogenital", "cardiovascular", "lymphatic", "nervous", "brain", "metabolism", "reproductive", "behavior", "sensory"))

# add dataset to characters
final$charlabels <- paste(final$charlabels, ' [', final$file, ' character ', final$charnums, ']', sep="")

# reorder based on character partitions
final <- final[, order(final$charpart)]

final

# and finally, plot:
plot(final, fill="charpartition")
```



## Find duplicates and put close together

```{r}
# find
dupmap <- read.csv("data/dups/dups.csv")
# dupmap$file1 <- paste(dupmap$dataset1, '.nex', sep='')
# dupmap$file2 <- paste(dupmap$dataset2, '.nex', sep='')


# match 1
id1 <- sapply(1:nrow(dupmap), function(x) {
  res <- final$file == dupmap$dataset1[x] & final$charnums==dupmap$char1[x]
  ifelse(any(res), which(res), NA)
  })

# match 2
id2 <- sapply(1:nrow(dupmap), function(x) {
  res <- final$file==dupmap$dataset2[x] & final$charnums==dupmap$char2[x]
  ifelse(any(res), which(res), NA)
  })

dupid <- na.omit(cbind(id1, id2))

verts <- sort(unique(c(dupid[,1], dupid[,2])))

edgs <- dupid

g <- graph_from_data_frame(edgs, vertices=verts, directed=FALSE)

# find groups of characters
grps <- max_cliques(g)
grps <- sapply(1:length(grps), function(x) {as.numeric(names(grps[[x]]))})
names(grps) <- paste('dup', seq_along(grps),sep='')

dup <- rep("", ncol(final$data))

com <- components(g)

dup[as.numeric(names(com$membership))] <- com$membership
dup[dup==""] <- NA
head(dup)

# final$charlabels <- paste(dup, final$charlabels, sep="")


# plot biggest and representative smallest (n=2) clique
pdf(file = "figure/dupclique_big.pdf")
plot(induced_subgraph(g, v=com$membership==which.max(com$csize)))
dev.off()

pdf(file = "figure/dupclique_small.pdf")
plot(induced_subgraph(g, v=com$membership==which.min(com$csize)))
dev.off()

# par(mfrow=c(4,6), mar=c(0,0,0,0), mex=.5, ps=6)
# for (i in seq_along(com$csize)) {
#   plot(induced_subgraph(g, v = which(com$membership==i)))
# }

df <- data.frame(char=as.numeric(names(com$membership)), group=com$membership)
df$charpart <- final$charpart[df$char]
dfsum <- df %>% group_by(group,charpart) %>% summarize(n=length(char)) %>% arrange(n)
dfsum$charpart <- factor(dfsum$charpart)
dfsum$charpart <- droplevels(dfsum$charpart)
dfsum$charpart <- factor(dfsum$charpart, levels = levels(dfsum$charpart)[rev(order(aggregate(n~charpart, dfsum, length)$n))])

ggplot(dfsum, aes(x = n)) +
  geom_histogram(binwidth = 1) +
  facet_wrap(~charpart) +
  xlim(2, NA) +
  # geom_vline(xintercept = c(2,3)) +
  labs(x = "Number of duplicate characters within clique", y = "Number of cliques") +
  theme_bw() +
  theme(strip.background=element_blank())

ggsave("figure/dupcliques.pdf")

# find screwed up (character paritions not the same):

# get character partitions for groups
x <- lapply(seq_along(com$csize), function(x) {
  final$charpartition[as.numeric(names(which(com$membership==x)))]
})

# find ones with > 1 trait category
x <- which(sapply(lapply(x, unique), length)>1)
length(x)
x

# reorder so duplicates are close together...

```




"Fate of characters" plot

```{r}
year <- str_extract(final$file, "\\d{4,4}$")

year[is.na(year)] <- "2016"

year <- as.numeric(year)

charpart <- final$charpart

df$year <- year[df$char]

df$group <- as.character(df$group)

sorting <- tapply(df$year, df$group, min)

df$group <- as.character(df$group)

df$group <- reorder(df$group, match(df$group, names(sorting)))

df$group <- factor(df$group, levels = df[order(df$year), "group"])

ggplot(data=NULL) +
  geom_point(data=df, aes(year, group, color = charpart)) +
  geom_path(data=df, aes(year, group, color = charpart)) +
  facet_wrap(~charpart, scales="free_y") +
  theme_bw() +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.grid = element_blank(), legend.position = "none")

ggsave("figure/character_fate.pdf", width=8, height=6)
```




## Output stuff to NEXUS file

```{r}
write.nex(final, file = "output/final_reordered.nex", format="nexus")
# write.nex(allnex.genome, 'output/palaeognath_genomes.nex')

test <- data.frame(char=names(com$membership), group=com$membership) %>% arrange(group) %>% dplyr::select(group, char)
test$char <- as.numeric(as.character(test$char))
test$charlabel <- final$charlabel[test$char]
write.csv(test, "output/duplicates.csv", row.names=FALSE)
```


## Make a supraspecific terminal tree

```{r}
# replace _spp with genus name
nms <- dinornithids
fixes <- rep("Dinornithidae", length(nms))
names(fixes) <- nms
map <- as.list(fixes)
map <- list("Dinornithidae"=dinornithids, "Aepyornithidae"=aepyornithids, "Tinamidae"=tinamous, "Lithornithidae"=lithornithids, "Dromaiidae"=emus, "Apterygidae"=apteryx, "Rheidae"=rhea, "Casuariidae"=casuarius, "Struthionidae"=struthio, "Anseriformes"=grep("Anas|Chauna", outgroups, value=T), "Galliformes"=grep("Megapodius|Gallus", outgroups, value=T), "Ichthyornis"=grep("Ichthyorn", outgroups, value=T))
map <- cbind(clade=rep(names(map), times=sapply(map, length)), taxon=unlist(map))
map <- as.list(setNames(map[,'clade'], as.character(map[,'taxon'])))

# collapse terminals
test <- collapse(final, map = map, method = "merge")

# subset to include desired clades
allnex_supra <- subset(test, taxlabels %in% c("Apterygidae", "Dinornithidae", "Aepyornithidae", "Tinamidae", "Lithornithidae", "Rheidae", "Casuariidae", "Struthionidae", "Anseriformes", "Galliformes", "Ichthyornis", "Dromaiidae"))

plot(allnex_supra)

table(is.na(allnex_supra$data))  # 9322/(20738+9322) -- 31% missing data compared to 

table(is.na(final$data)) # 271886/(271886+41239) = 87% missing data

# write.nex(allnex_supra, file="analyses/mrbayes run2/testsupra.nex", mrbayes=TRUE, ngen=1e6, run=FALSE)

# Would it be useful to visualize the number of polymorphic characters by body region?

# Number of polymorphic characters
length(grep("\\(", allnex2$data))
length(grep("\\(", allnex_supra$data))
```


## Run TNT analyses

Make separate files for different character sets to run in TNT

```{r tntrun, eval=FALSE}
# take out duplicates
allnex_supra_nodup <- allnex_supra[, -as.numeric(names(com$membership))]

# make Ichthyornis first (outgroup in TNT analyses is first row)
id <- grep("Ichthy", allnex_supra_nodup$tax)
allnex_supra_nodup <- allnex_supra_nodup[c(id, setdiff(1:nrow(allnex_supra_nodup$data), id)), ]
plot(allnex_supra_nodup)

# output in TNT format to specified folders by named by character partition
# run analyses in terminal using: tnt 'proc file.tnt; exactsol.run; zzz'
oldwd <- getwd()
for (i in unique(allnex_supra_nodup$charpartition)) {
      x <- subset(allnex_supra_nodup, charpartition==i)
      system(paste0("mkdir analyses/tnt2/", i, ""))
      write.nex(x, file=paste0("/Users/chadeliason/Documents/UT/projects/phenome/analyses/tnt2/", i, "/", i, ".tnt"), format="tnt")
      system(paste0("cp analyses/tnt2/exactsol.run analyses/tnt2/", i))
      setwd(paste0("/Users/chadeliason/Documents/UT/projects/phenome/analyses/tnt2/", i))
      system(paste0("tnt 'proc ", i, ".tnt; exactsol.run; zzz;'"))
}
setwd(oldwd)
# note - there is probably a better way to do this...but at least i shouldn't have to keep coming back to start a new analysis
```

Plot resulting trees

```{r tntresults}
treelist <- list.files(path="analyses/2015-08-28 tnt2", pattern="*.tre", include.dirs=TRUE, full.names=TRUE, recursive=TRUE)

# read in trees in subfolders
trees <- lapply(treelist, function(x) {
      # removing white space in tree files
      res <- readLines(x)
      res <- gsub(" \\;", "\\;", res)
      id <- grep("begin", res)
      cat(res[1:id], "translate", paste0(1:12, " ", allnex_supra_nodup$tax, collapse=",\n"), ";", res[(id+1):length(res)], sep="\n", file="ex.tre")
      # read in tree
      # cat(x, file="ex.tre", sep="\n")
      read.nexus("ex.tre")
      # assign tips
      # plot
      }
)

id <- which(sapply(trees, class)=="multiPhylo")

for (i in id) {
      trees[[i]] <- consensus(trees[[i]])
}

sapply(trees, class)

names(trees) <- str_match(treelist, "([a-z]+)\\/ratites\\.tre")[,2]

pdf(file="~/Dropbox/Chad/pars_impenum_cons_trees.pdf", width=8, height=5)
par(mfrow=c(2,4), mar=c(1,1,1,1))
lapply(seq_along(trees), function(x) plot(ladderize(trees[[x]]), main=names(trees)[[x]]))
dev.off()
```


## Summarize the concatenated dataset

```{r}
# reorder data matrix
tmp <- allnex2

# rowsort <- rev(order(sapply(1:nrow(tmp$data), function(x) {sum(!is.na(tmp$data[x,]))})))
# colsort <- rev(order(sapply(1:ncol(tmp$data), function(x) {sum(!is.na(tmp$data[,x]))})))
# tmp <- tmp[rowsort, colsort]

# tmp <- allnex.genome[match(c(ingenome, fossils, outgroups), allnex.genome$tax), ]

# tmp$taxlabels[which(tmp$taxlabels %in% fossils)] <- paste0('â€ ', tmp$taxlabels[which(tmp$taxlabels %in% fossils)])
tmp$taxlabels[which(tmp$taxlabels %in% ingenome)] <- paste0('*', tmp$taxlabels[which(tmp$taxlabels %in% ingenome)])

# plot(allnex.genome, bw=T)

# how organize by completeness of data?

table(is.na(allnex2$data))/length(allnex2$data)  # ~75% missing data :( - more when we include the two other kiwis (maybe we can just use Apteryx australis - which livezey scored?) 

# df <- as.data.frame(table(allnex.genome$file))

df <- as.data.frame(table(allnex2$file, allnex2$charpartition))
df$type <- as.character(df$Var2)
df$type[grep('chromosomes|digestive|nervous|respiratory', df$type)] <- "Other"
df$type <- factor(df$type)
ggplot(df, aes(reorder(Var1, Freq), Freq, fill=type)) +
      geom_bar(stat='identity') +
      labs(x='Dataset', y='Number of variable characters') +
      # coord_fixed(1/50, ylim=c(0, 1000)) +
      ylim(0,1000) +
      coord_flip() +
      theme_bw() +
      theme(aspect.ratio=1.5)

# df <- data.frame(file=allnex2$file, charpart=allnex2$charpart)
# df$charpart <- as.character(df$charpart)
# df$charpart[grep('chromosomes|digestive|nervous|respiratory', df$charpart)] <- "Other"
# df$charpart <- factor(df$charpart)
# head(df)
# ggplot(df, aes(x=file,fill=charpart)) + geom_histogram() + coord_flip()

# ggplot(df, aes(reorder(Var1, Freq), Freq, fill=type)) +
#       geom_bar(stat='identity') +
#       labs(x='Dataset', y='Number of variable palaeognath characters') +
#       coord_flip() +
#       theme_bw()

df2 <- df %>% group_by(Var2) %>% summarize(Freq=sum(Freq))
ggplot(df2, aes(reorder(Var2, Freq), Freq, fill=type)) +
      geom_bar(stat='identity') +
      labs(x='System', y='Number of variable characters') +
      ylim(0, 1000) +
      coord_flip() +
      theme_bw() +
      theme(aspect.ratio=2)

# table of dataset by # of characeter types (by anatomical region)

xtabs(~allnex2$file+allnex2$charpart, sparse=T)
# image(xtabs(~allnex.genome$file+allnex.genome$charpart, sparse=T), col.region=heat.colors(25))


## Data matrix connectivity analysis

# create bipartite network graph

# find taxa without all NAs for each dataset (e.g., file)
# out <- matrix(nrow=length(tmp$taxlabels), ncol=length(unique(tmp$file)))
# dimnames(out) <- list(tmp$taxlabels, unique(tmp$file))
# for (i in seq_along(rownames(out))) {
#       for (j in seq_along(colnames(out))) {
#             dat <- tmp[tmp$taxlabels==rownames(out)[i], tmp$file==colnames(out)[j]]
#             # print(dat)
#             out[i, j] <- as.numeric(!all(is.na(dat$data)))
#       }
# }

# M <- matrix(as.numeric(is.na(allnex.reordered$data)), nrow=nrow(allnex.reordered$data))

# g <- graph_from_incidence_matrix(M[1:10, 1:10])

# is_bipartite(g)

# plot(g, layout=layout.bipartite)

# largest_cliques(g)

# connectivity among taxa in scored character sets
# image(t(out), col=rev(grey.colors(2)))
# rowsort <- order(apply(out, 1, sum))
# colsort <- rev(order(apply(out, 2, sum)))
# image(t(out[rowsort, colsort]), col=rev(gray.colors(2)))


connectivity <- function(x, ...) {
  require(igraph)
  res <- x$data
  res <- ifelse(is.na(res), 0, 1)
  dimnames(res) <- list(x$taxlabels, x$charlabels)
  g <- graph.incidence(res, weighted=TRUE)
  obs.all <- bipartite.projection(g)[[1]]
  # E(obs.all)$curved <- 0.2
  plot.igraph(obs.all, edge.width=E(obs.all)$weight/ncol(res)*10, ...)
}


# connectivity(tmp, vertex.size=5, vertex.color='red')

# remove taxa without any scorings
# tmp <- allnex2[sapply(1:nrow(allnex2$data), function(x) {!all(is.na(allnex2$data[x, ]))}), ]
tmp <- allnex.genome[sapply(1:nrow(allnex.genome$data), function(x) {!all(is.na(allnex.genome$data[x, ]))}), ]

pdf(file="figs/connectome2.pdf")
par(mar=c(0,0,0,0))
connectivity(tmp, vertex.label.cex=0.5, vertex.color=clades$clade[match(tmp$tax, clades$spp)], vertex.size=5, edge.color=gray(.5, alpha=.5))
dev.off()


# connectivity(tmp, vertex.label=NA)
# connectivity(allnex.genome, vertex.size=3, vertex.label.cex=0.5)

# source("~/Desktop/computeBicliques.R")
# maybe plot a network with number of characters in common between different taxa?
# res <- computeBicliques(g,l=17,k=3)  # l = # of 'rows', k = number of vertices connected to edge (2 for bipartite?)
# res
```


## Grep matching for morphology subsets

```{r, eval=FALSE}
# other subsets
vomer <- subset(allnex.genome, grepl('[Vv]omer', charlabels))
vomer <- na.omit(vomer)
plot(vomer, bw=T)

skull <- na.omit(subset(allnex.genome, charpartition=='skull'))
plot(skull, fill="file", legend.pos = "top")

hindlimb <- na.omit(subset(allnex.genome, charpartition=='hindlimb'))
plot(hindlimb, fill="file", legend.pos = "top")
```




## Test find duplicates function

```{r, eval=FALSE}
x <- read.nex('example/testdup.nex')

grid.arrange(plot(x), plot(duplicated(x, map=list(c(1,2), c(4,6)))), nrow = 2)


# now, we need to find the duplicates!!

# method 1: find duplicates based on comments in character sets

# find where char. is within NEXUS comment, i.e. '[ ]'
ids <- grep('\\[.+[Ch]ar[\\.]*.+\\]', allnex3$statelabels)
allnex3$file[ids]
allnex3$statelabels[ids]
# cat(paste0(ids,' ',allnex3$statelabels[ids]), sep='\n\n', file='~/Desktop/crap.txt')
cat(paste0(ids,' ',allnex3$statelabels[ids]), sep='\n\n')

# find patterns like:
# Worthy and Holdaway 2002: appendix 3, char. 72 -- keep 'Worthy and Holdaway 2002' and '72'
# Bourdon et al. (2009a: char. 126) -- keep 'Bourdon et al. (2009)' and '126'
# Worthy & Holdaway (2002: char. 67) -- keep 'Worthy & Holdaway (2002)' and '67'
# Bourdon et al. (2009a: chars. 121, 123). -- keep 'Bourdon et al. (2009)' and '121' '123'

grep('Worthy\\s(&|and)\\sHoldaway\\s[\\(]?\\d+', allnex3$statelabels, value=TRUE, perl=TRUE)

str_match(allnex3$statelabels, regex('(Worthy\\s(?:&|and)\\sHoldaway)\\s[\\(]?(\\d+)', perl=TRUE))

grep('Bourdon(?:\\set\\sal\\.\\s)[\\(]?\\d{4}[\\)]?', allnex3$statelabels, value=TRUE, perl=TRUE)

str_match(allnex3$statelabels, regex('([A-Z][a-z]+\\s(?:&|and)(?:\\s[A-Z][a-z]+\\s))[\\(]?(\\d{4})[\\)]', perl=TRUE))

# match character number
str_match(allnex3$statelabels, '[Ch]ar[\\.]?\\s(\\d+)')

str_match('Bourdon et al. (2009a: character 122)', '[Ch]ar(?:\\.|acter)\\s(\\d+)')

str_match('Bourdon et al. (2009a: character 122)', '(Bourdon)(?:\\set\\sal\\.\\s)[\\(]?(\\d{4})[\\)]?[a-z]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')

str_match('Bourdon et al. (2009a: char. 122)', '([A-Z][a-z]+)(?:\\set\\sal\\.\\s)?[\\(]?(\\d{4})[\\)]?[a-z]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')

str_match('Worthy & Scofield 2009 character 122', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s))[\\(]?(\\d{4})[\\)]?[a-z]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')

str_match('Worthy (2009) character 122', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')

str_match('Worthy 2009 character 122', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')

# test the final product
# finds cases within text of '...AUTHOR YEAR CHARACTER XX...'
str_match('Worthy 2009 char. 122', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')
str_match('Worthy and Holdaway 2002: appendix 3, char. 72', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')
str_match('Boudon et al. (2009a: char. 126)', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')
str_match('Worthy & Holdaway (2002: char. 67)', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.+[Ch]ar(?:\\.|acter)\\s(\\d+)')

# only find closeby characters
str_match('Worthy & Holdaway (2002: char. 67)', '([A-Z][a-z]+\\s(?:et\\sal\\.\\s|(?:\\&|and)\\s[A-Z][a-z]+\\s)?)[\\(]?(\\d{4})[\\)]?.{1,5}[Ch]ar(?:\\.|acter)\\s(\\d+)')



# test find duplicate function

skull <- subset(allnex.genome, charpartition=="skull")

plot(skull)

skull.fixdup <- duplicated.nex(x = skull)




# output new file

write.nex(allnex3, 'output/concat+merge.nex')

```


## Test buildnex function (reading PDF and converting to a nexus object)

```{r, eval=FALSE}
tmp1 <- buildnex(pdffile = 'example/Bertelli_Chiappe_2005.pdf', ntax = 34, nchar = 63, first = 20, last = 22)
# plot(tmp1, legend.pos = 'right')

tmp2 <- buildnex(pdffile = 'example/Johnston_2011.pdf', ntax = 13, nchar = 57, first = 21, last = 24)
# plot(tmp2, legend.pos = 'right')

tmp3 <- buildnex(pdffile = 'example/Bertelli_2014.pdf', ntax = 56, nchar = 157, first = 24, last = 25)
# tmp3 <- sort(tmp3)
# plot(tmp3, legend.pos = 'right')

tmp4 <- buildnex(pdffile = 'example/Mayr_2003.pdf', ntax = 46, nchar = 148, first = 20, last = 24)
# plot(tmp4, legend.pos = 'right')

# use character categories in baker et al. 2014?
# skull
# wing
# pectoral girdle
# pelvic girdle
# leg & foot
# vertebrae, trachea & thyroid(?) bone
```


## Plotting integumentary characters with phylogeny

```{r}
tre <- read.nexus("/Users/chadeliason/Documents/UT/projects/phylogenetic trees/birds/mitchell 2014/DataFiles+Trees/2_MolecularDating/2.1_BEAST_AllTaxa/Mitochondrial_AllTaxa.tree")

tre <- extract.clade(tre, getMRCA(tre, c("Struthio_camelus", "Tinamus_major")))

integument <- na.omit(subset(allnex.genome, charpartition == 'integument'))

plot(integument, phy = tre, bw = FALSE, fsize = 8)

# urogenital <- na.omit(subset(allnex.genome, charpartition == 'urogenital'))
# plot(urogenital, phy = tre, bw = FALSE, fsize = 8)

# forelimb <- na.omit(subset(allnex.genome, charpartition == 'forelimb'))
# plot(forelimb, phy = tre, bw = FALSE, fsize = 8)
```



## find and fix(?) `bad' characters

```{r, eval=FALSE}
# find characters with absent + multistate

ids <- grep('absent', allnex3$state)

ids

setNames(sapply(allnex3$statelabels, length), NULL)

paste0(allnex3$file[ids], ' ', allnex3$charnum[ids])
```


## Create subset data matrix

```{r}
ss <- 'Apteryx_austr|Apteryx_oweni|Eudromia_elegans$|Chauna|Struthio_|Rhea|Dromaius|Casuarius_casuarius|Casuarius_bennetti|Nothura_maculosa|Pterocnemia|Crypturellus_soui|Crypturellus_obsoletus|Anseranas|Anas|Megapodius|Gallus$'

# ss <- 'Apteryx_austr|Apteryx_oweni|Eudromia_elegans$|Struthio|Rhea|Dromaius|Casuarius_casuarius|Tinamus_major|Nothura_maculosa|Pterocnemia|Crypturellus_soui'
# livezey <- nexs[[grep('livezey_2006.nex', files)]]
# tmp <- allnex2[grepl(ss, allnex2$taxlabels), allnex2$file=='livezey_2006.nex']  # test_livezey
# tmp <- livezey[grepl(ss, livezey$taxlabels), ]  # test_livezey
tmp <- allnex2[grepl(ss, allnex2$taxlabels), ]
# tmp <- subset(allnex2, grepl(ss, taxlabels))
rowsort <- rev(order(sapply(1:nrow(tmp$data), function(x) {sum(!is.na(tmp$data[x,]))})))
colsort <- rev(order(sapply(1:ncol(tmp$data), function(x) {sum(!is.na(tmp$data[,x]))})))
tmp <- tmp[rowsort, colsort]
# tmp <- na.omit(tmp)
plot(tmp)
```




## Run preliminary MrBayes analysis

```{r, eval=FALSE}
# I setup constraints manually in the .nex file. is there a way to do this automatically? or maybe send it to mrbayes given an input in the write.nex() function?

# TODO - set this up so you can automatically get charsets from those defined in nexus R object

# tmp <- subset(allnex2, grepl('Aepyornis|Lithornis_spp|Gallus_gallus$|Chauna|Struthio|Rhea|Ichthyornis|Apsara|Eudromia_elegans$|Dromaius|Apteryx_austr|Casuarius_casuarius|Tinamus_major|Dinornis_spp|Palaeotis|Gavia', taxlabels))

# plot(tmp)

# order by character partition
# tmp <- tmp[, order(tmp$charpart)]

# find 1st of each unique character partition
# ids <- which(duplicated(tmp$charpart)==0)
# nmax <- ncol(tmp$data)
# ids2 <- c((ids[-1]-1), nmax)
# cat(paste0('charset ', unique(tmp$charpart), '=', ids, '-', ids2, ';'), sep="\n")

write.nex(tmp, file='~/Desktop/test/test.nex', mrbayes=TRUE, run=FALSE, phy=NULL, ngen=1000000)
# write.nex(tmp, file='~/Desktop/test_livezey/test.nex', mrbayes=TRUE, run=FALSE, phy=NULL, ngen=1000000)
```


Diagnose convergence

```{r}
x <- load.multi('~/Desktop/ratites_supraspec_mrbays_2015_08_28')

makeplot.treespace(x, n.points=25, burnin=100)

compare.n(x, burnin=10)

makeplot.param(x)

makeplot.posteriors(x, burnin=10, window.num=10)
```


Plot phylogeny with uncertainty

```{r}
trees <- c(x[[1]]$trees, x[[2]]$trees)
# trees2 <- lapply(trees, root, outgroup='Crocodylia', resolve.root=TRUE)
# trees2 <- lapply(trees, drop.tip, tip='Crocodylia')
trees2 <- lapply(trees, compute.brlen)
class(trees2) <- 'multiPhylo'
par(mar=c(3,0,0,2))
densiTree(trees2[sample(1:length(trees2), 500, replace=F)], col=1)

tree <- read.nexus("~/Desktop/test/test.nex.con.tre")

pdf(file="output/mrbayes_palaeognath_mcc_tree.pdf", width=7, height=5)
par(mar=c(0,0,0,0))
plot(tree, type='u', cex=0.5, lab4ut="axial", label.offset=.04)
dev.off()
```


Find synapomorphies of ratites

```{r}
x <- grep('Dromaius|Struthio|Rhea|Apteryx_oweni', allnex.genome$tax)
y <- grep('Eudromia_elegans$', allnex.genome$tax)

ss <- allnex.genome[x, ]$data

# share same state
id1 <- sapply(1:ncol(ss), function(x) {length(unique(na.omit(ss[,x])))})==1
# different from tinamous
id2 <- sapply(1:ncol(ss), function(x) {!(any(unique(na.omit(ss[,x])) %in% allnex.genome[y, x]$data))})

plot(allnex.genome[x, id1 & id2])

ss <- allnex.genome[x, id1 & id2]

cat(sort(ss$charlab), sep="\n")

sort(table(ss[, order(ss$charpartition)]$charpart))

na.omit(subset(ss, charpartition=="integument"))$data
```



